model/src/main/java/org/activityinfo/model/util/HeapsortColumn.java

```diff
@@ -19,186 +19,208 @@
 
 package org.activityinfo.model.util;
 
+import com.google.common.primitives.UnsignedBytes;
+import com.google.gwt.core.shared.GwtIncompatible;
+import it.unimi.dsi.fastutil.ints.Int2DoubleOpenHashMap;
+import org.activityinfo.model.query.ColumnView;
+
+import java.util.BitSet;
+
 /**
  * This utility heapsorts a column, reordering the given index arrays but leaving the value arrays unchanged.
  * Also allows for selection of a specific row range to sort on.
- * Derived from the {@link HeapsortTandem} class.
  */
 public final class HeapsortColumn {
 
-    private static int compare(String str1, String str2) {
-        if (str1 == str2) {
-            return 0;
-        } else if (str1 == null) {
-            return -1;
-        } else if (str2 == null) {
-            return 1;
+    private static boolean isLessThan(double x, double y, boolean ascending) {
+        if(ascending) {
+            if (x < y) {
+                return true;
+            }
+            if (Double.isNaN(x) && !Double.isNaN(y)) {
+                return true;
+            }
+            return false;
+
         } else {
-            return str1.compareTo(str2);
+            if(y < x) {
+                return true;
+            }
+            if (Double.isNaN(y) && !Double.isNaN(x)) {
+                return true;
+            }
+            return false;
         }
     }
 
-    private static int compare(String[] labels, int enum1, int enum2) {
-        if (enum1 == enum2) {
-            return 0;
-        } else if (enum1 == -1) {
-            return -1;
-        } else if (enum2 == -1) {
-            return 1;
+    /**
+     * Given two numbers encoded using
+     * @param bx
+     * @param by
+     * @param ascending
+     * @return
+     */
+    @GwtIncompatible
+    private static boolean isLessThan(byte bx, byte by, boolean ascending) {
+        // Treat as unsigned
+        int x = UnsignedBytes.toInt(bx);
+        int y = UnsignedBytes.toInt(by);
+        if(ascending) {
+            if(x == 0 && y != 0) {
+                return true;
+            }
+            return UnsignedBytes.toInt(bx) < UnsignedBytes.toInt(by);
+
         } else {
-            return labels[enum1].compareTo(labels[enum2]);
+            if(y == 0 && x != 0) {
+                return true;
+            }
+            return y < x;
         }
     }
 
-    private static int compare(String[] labels, byte enum1, byte enum2) {
-        if (enum1 == enum2) {
-            return 0;
-        } else if (enum1 == -1) {
-            return -1;
-        } else if (enum2 == -1) {
-            return 1;
+    private static boolean isLessThan(short sx, short sy, boolean ascending) {
+        // Treat as unsigned - promote to int with a 16-bit mask
+        int x = sx & 0xFFFF;
+        int y = sy & 0xFFFF;
+        if(ascending) {
+            if(x == 0 && y != 0) {
```

Comment:
This probably would benefit from some commentary... we will both forget what this line does in 6 months. One way to assign the expressions to variables to give them names...

```
// missing values encoded as zeros
boolean xMissing = (x == 0);
boolean yMissing = (y == 0);
```

Reply:
Will do